### PR TITLE
Add support for .NET8.0 HttpClient metrics

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Towards https://github.com/open-telemetry/opentelemetry-dotnet/issues/4799

Details: https://github.com/open-telemetry/opentelemetry-dotnet/issues/4799#issuecomment-1743421346
https://github.com/open-telemetry/opentelemetry-dotnet/issues/4799#issuecomment-1743435764

## Changes

This PR adds support for metrics enablement that are part of `.NET8.0`. For `HttpClient` the following meters will be added to the MeterProviderBuilder when application targets `.NET8.0` framework:

 * **Meter** : `System.Net.Http`
    * `http.client.request.duration`
    * `http.client.active_requests`
    * `http.client.open_connections`
    * `http.client.connection.duration`
    * `http.client.request.time_in_queue`

  * **Meter** : `System.Net.NameResolution`
    * `dns.lookups.duration`

Users may selectively opt-out with the help of views.

**TODO**: Update Readme

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs
Comment (utpilla) [inline test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs:177]: Why is this needed?

File: test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs
Comment (utpilla) [inline test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs:679]: Use the property `Count` instead?

File: test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs
Comment (utpilla) [inline test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs:775]: Extract these bounds into a variable.

File: test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs
Comment (vishweshbankwar) [reply test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs:177]: THis was related to https://github.com/open-telemetry/opentelemetry-dotnet/issues/4928. I have fixed this now.

File: test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs
Comment (vishweshbankwar) [reply test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs]: This is now removed.

File: test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs
Comment (vishweshbankwar) [reply test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs]: Same as https://github.com/open-telemetry/opentelemetry-dotnet/pull/4931#discussion_r1352858883

**Comment by samsp-msft** [comment]
@JamesNK please review

File: src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md:38]:
```suggestion
  `.NET8.0` framework or newer.
```

File: src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md:40]: as an end-user reading this, its not obvious what are these metrics. Its best to clarify that some of the metrics are part of the OTel Semantic Conventions (link to it). 

Also - does these metrics respect the OTEL_SEMCONV_STABILITY_OPT_IN flag? Please clarify that briefly in the changelog. (And of course Readme.md should have full details)

Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md:40]: Ideally, we would link all metrics to this https://github.com/open-telemetry/semantic-conventions/pull/283. We need to wait for that PR to merge first.

I will clarify the flag part.

File: src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md
Comment (antonfirsov) [inline src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md:48]: Heads up that this will be renamed to `dns.lookup.duration` in .NET 8.0 GA!

File: test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs
Comment (antonfirsov) [inline test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs:202]: Failure expected with the GA update in november.

File: src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md
Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md:48]: Thanks @antonfirsov. We will update the doc accordingly then.

File: src/OpenTelemetry.Instrumentation.Http/MeterProviderBuilderExtensions.cs
Comment (sshumakov) [inline src/OpenTelemetry.Instrumentation.Http/MeterProviderBuilderExtensions.cs:67]: I think I found code enabling `http.client.request.duration` instrumentation mentioned in #4932.

@antonfirsov mentioned the following:

> Note that HttpMetricsEnrichmentContext is a low-level primitive that needs a callback registration call for each HttpRequestMessage, and the assumption was that higher-level libraries will build their own (more convenient) enrichment experiences around it. 

I am trying to understand - are you planning to add a feature to open telemetry, supporting http client request enrichment? If this is not something planned to be part of open telemetry, then is expectation that enrichments will be attached using custom http client handler, with code passing enriching callback?

```csharp
HttpMetricsEnrichmentContext.AddCallback(message, (ctx) =>
    {
        ctx.AddCustomTag("MyCustomTag", "MyCustomValue");
    });
```

@vishweshbankwar

Comment (antonfirsov) [reply src/OpenTelemetry.Instrumentation.Http/MeterProviderBuilderExtensions.cs:67]: > I think I found code enabling http.client.request.duration instrumentation mentioned in #4932.

I would assume this alone will not enable the instruments, and you would also need `AddInMemoryExporter`. @vishweshbankwar how does this work?

Comment (antonfirsov) [reply src/OpenTelemetry.Instrumentation.Http/MeterProviderBuilderExtensions.cs:67]: > I am trying to understand - are you planning to add a feature to open telemetry, supporting http client request enrichment?

I think it would be the best if `opentelemetry-dotnet` exposed an `HttpMetricsEnrichmentContext`-based replacement of their current enrichment API-s:

https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Instrumentation.Http#enrich-httpclient-api

Designing a global callback-registration API is an opinionated area. It depends on how users manage their HttpClients. (ie. an API on HttpClient doesn't help much, if you use IHttpClientFactory.)

@JamesNK any thoughts?

Comment (sshumakov) [reply src/OpenTelemetry.Instrumentation.Http/MeterProviderBuilderExtensions.cs:67]: We also need to think about cases, when http client is used by referenced package, which does not expose it, but we still want to collect these metrics and attach custom tags. Not sure if this feature is available.

Comment (JamesNK) [reply src/OpenTelemetry.Instrumentation.Http/MeterProviderBuilderExtensions.cs:67]: What are you trying to achieve? What are the scenarios?

If someone has configured a callback to a `HttpRequestMessage` with `HttpMetricsEnrichmentContext.AddCallback`, then the tags added should automatically flow through into the `http.client.request.duration` counter. There is no need to add features to OpenTelemetry.

`HttpMetricsEnrichmentContext.AddCallback` is a low-level API but dotnet/extensions is adding support for configuring enrichers via DI.

File: src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md
Comment (utpilla) [inline src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md:51]: Same as https://github.com/open-telemetry/opentelemetry-dotnet/pull/4934#discussion_r1362896717

File: test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs
Comment (utpilla) [reply test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs:177]: So is this needed?

File: src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md
Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md:40]: Clarified the flag part. PTAL. The PR is still not merged, will need to come back and update the links after the merge.

File: src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs
Comment (utpilla) [inline src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs:49]: Remove this?

File: src/OpenTelemetry.Instrumentation.Http/MeterProviderBuilderExtensions.cs
Comment (utpilla) [inline src/OpenTelemetry.Instrumentation.Http/MeterProviderBuilderExtensions.cs:75]: nit: Why change the indentation?

File: src/OpenTelemetry.Instrumentation.Http/MeterProviderBuilderExtensions.cs
Comment (utpilla) [inline src/OpenTelemetry.Instrumentation.Http/MeterProviderBuilderExtensions.cs:68]: Could we move this check to the top of the method and return the builder after adding the meters if it's on net8.0 or greater?

File: test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs
Comment (utpilla) [inline test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs:194]: nit: Use `Count()` instead of `Where()`

Comment (utpilla) [reply test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs:194]: nit: A stronger test would be to individually look for each expected metric instead of looking at the aggregated count.

**Review by utpilla — APPROVED** [review]

File: src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md:51]:
```suggestion
  * When targeting `.NET8.0` framework or newer, `http.client.request.duration` metric
```

**Review by cijothomas — APPROVED** [review]
Looks good.
Some comments from https://github.com/open-telemetry/opentelemetry-dotnet/pull/4934#pullrequestreview-1689154215 are applicable here as well.

File: src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md
Comment (CodeBlanch) [reply src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md:40]: One suggestion...

@noahfalk shared this doc with me: https://learn.microsoft.com/dotnet/core/diagnostics/built-in-metrics

We should probably link to it here and in aspnetcore instrumentation?

File: src/OpenTelemetry.Instrumentation.Http/MeterProviderBuilderExtensions.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.Http/MeterProviderBuilderExtensions.cs:46]: nit: Indention looks a bit odd.

```suggestion
        return builder
            .AddMeter("System.Net.Http")
            .AddMeter("System.Net.NameResolution");
```

File: test/OpenTelemetry.Instrumentation.Http.Tests/HttpTestData.cs
Comment (CodeBlanch) [inline test/OpenTelemetry.Instrumentation.Http.Tests/HttpTestData.cs:55]: This is kind of interesting. How is it we don't report 2.0 now?

**Review by CodeBlanch — APPROVED** [review]
Some non-blocking feedback, LGTM

File: src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md
Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md:40]: This is great. I will fix the links. I was not aware of this one., was waiting on spec PR to merge. Thanks!

File: test/OpenTelemetry.Instrumentation.Http.Tests/HttpTestData.cs
Comment (vishweshbankwar) [reply test/OpenTelemetry.Instrumentation.Http.Tests/HttpTestData.cs:55]: @CodeBlanch See https://github.com/open-telemetry/opentelemetry-dotnet/issues/4928

File: test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs
Comment (vishweshbankwar) [reply test/OpenTelemetry.Instrumentation.Http.Tests/HttpClientTests.cs:194]: good point. added for follow up here https://github.com/open-telemetry/opentelemetry-dotnet/issues/4972

**Pull request merged**
